### ISSUE TITLE
Linux Client need modification

### ISSUE THREAD

**Issue by tioui** [ArchipelaGOAL/Archipelago#5]
The Archipelago Client use the Windows only Pymem library to read location check. If we want a Linux (or other Posix system) port, we will need to replace it with another thing. Here three scenarios I propose:

- Find a multiplatform (cross os) library that works the same on Windows and on Linux.
    * I search a little bit on the internet and the only thing that I find is this: https://github.com/qb-0/PymemLinux but it has to be launched as root. I think that we should make this a last resort option.
- Change a little bit the client so that it starts the game (`gk` executable) himself and use Python-ptrace in Linux to read the memory instead of Pymem.
    * The problem with that option is that we will need to do the same code twice: on with Pymem for Windows and another with Python-ptrace for Linux.
- At last, changing the client structure to use only the repl connection (using the Web API) to read and write in the game variable. That way, The Pymem is not needed anymore to read in the game process memory.

**Comment by massimilianodelliubaldini** [comment]
- [ ] Use subprocess calls to `ps` to find the `gk` and `goalc` processes to ensure they don't die on us (rather than pymem)
- [ ] Use the REPL to query the `*ap-jak1-info*` memory structure and parse for your progress.
    - [ ] Prerequisite: Would need to modify the the REPL server itself to respond to requests properly, and with actual data not just "OK!"

**Comment by massimilianodelliubaldini** [comment]
It might actually be easier to just put in platform-specific python modules - Pymem for Windows, ptrace for Linux. I'm not against it as an option.

**Comment by massimilianodelliubaldini** [comment]
Though this kind of scares me:

> [This project is no longer maintained and is looking for a new maintainer.](https://python-ptrace.readthedocs.io/en/latest/)

**Comment by tioui** [comment]
It had an update last march. https://github.com/vstinner/python-ptrace/commits/main/

**Comment by tioui** [comment]
Also, we could probably use https://proc.readthedocs.io/en/latest/api.html#module-proc to know if a process existed instead of the subcommand hack.

**Comment by tioui** [comment]
>     * [ ]  Use the REPL to query the `*ap-jak1-info*` memory structure and parse for your progress.


A question: What is "ap-info-jak1". I don't find that in the source code.

**Comment by Minimistro32** [comment]
I can see there is help wanted here. I'm curious, approximately how much code is using Pymem and would need to get a Linux version using Pytrace? I don't have a lot of spare time, but I do have python and some low-level programming experience. 

From what I can tell at a quick glance: MemoryReader.py, ReplClient.py, and Client.py are the affected files. That sound right? No promises yet, just trying to scope out what size a project linux support might be.

**Comment by tioui** [comment]
I did look at the beginning of winter, but [massimilianodelliubaldini](https://github.com/massimilianodelliubaldini) was actually working hard on it and was doing a lot of modification really fast. So, I decided to pause a little on this project to work on my other project. I think that the code is quite stable at the moment, so it could be a good time to think about it again.

The last time I looked at it, there was modifications to do in almost every files of the client. A lot of hard-coded ".exe" and the use of power shell that I think is not necessary under Linux.

One of the problems is that the processes are a lot more secure from other processes (of what I know) and if on Windows, you can just start looking at process memory, it is not possible in Linux if you don't have root access or if the process looking is not the parent process of the process that is looked. So, this is a challenge too, it sill is needed to change the starting process.

Another problem is to be sure that the memory structure of the game process is the same as when compiled on Windows. If it is the case, it will probably be easier to create a routine that manages the memory reader for Windows and another one for Linux. If the process memory is not the same between Windows and Linux, we will have to change every place in the code that is reading some code or more likely since a structure is used in the game code, using relative addressing with a different base address.

In other words, lots of think to change and test.

**Comment by Minimistro32** [comment]
I see that is a lot to consider. Respectfully might be more than I can handle right now, especially if the memory structure is actually different like you mentioned.

That said, have you seen [PyMemoryEditor](https://github.com/JeanExtreme002/PyMemoryEditor)? I saw it mentioned in [this Pymem discussion post](https://github.com/srounet/Pymem/discussions/122). At a glance, it has Linux and Windows support for reading, writing, and searching memory. I'm nowhere familiar enough with this project to say. Do you think it could be robust enough to be used here?

I will say, PyMemoryEditor has the same caveat as Python-ptrace: it needs to be the parent process in order to function. But it has the upsides of being cross platform, actively maintained (near as I can tell), and would result in one codebase which can be a plus.

**Comment by tioui** [comment]
I did not know PyMemoryEditor. It is absolutely something we should be looking into. I think that using multi-platform tools are always better than using multiple separate tools for every platform. But since that impact the Windows build as well as the future Linux one, @massimilianodelliubaldini should accept that we change the Pymem usage. I would not like to work on something that would not be accepted in the repo.

As for the alternative to fixing REPL, I think that the best thing that could be done is actually to fix the REPL server to accept query as well as command. This would be better for REPL and better for the AP GOAL clients. The problem is, I looked at it and we are talking about massive modifications since the entire query structure are done with prints.

**Comment by jacobmix** [comment]
To update Linux people watching this issue. Markus made a post in Discord talking about the future of Linux support:
https://discord.com/channels/731205301247803413/1138467653552451606/1297923693783154720
> Since this thread has slowed down considerably, I'll just give a quick update on the near future plans for the mod. In this order, here are the tasks remaining:
> 
> 1. Fix an oversight where you can still do citadel skip V2 (already coded). 
2. Merge in the compiler change so we don't need to copy iso_data anymore (which may be a little bit of waiting).
3. Reopen discussion about merging Jak into Archipelago mainline.
> 
> ...That's it really. It's been feeling like the homestretch for a while, so I'm only just now catching up to the feeling.
> 
> For the Linux afficionados who are, miraculously, still around: I have not forgotten about you. But in the process of cleaning up a lot of infrastructural headaches, I learned that mods on OpenGOAL are not as well supportive of Linux as the vanilla game. That remains somewhat true even now, as opening the REPL specifically for mods doesn't work right. The 2nd task is going to fix one aspect of that, but it won't fix everything. So Linux support is - frustratingly I'm sure - just out of reach for the time being. 
> 
> Put it this way: if you were worried about Pymem being a Windows-only python module, don't be. That should be the literal last thing that needs to get tackled to support Linux. So I would be patient for now; a few things are out of all of our control for the time being.

**Comment by massimilianodelliubaldini** [comment]
To elaborate on a few things:

- OpenGOAL's very talented team is currently focused on getting Jak 3 working.
- _My_ current focus is on getting merged into Archipelago mainline.
-  AP's core developers' current focus is getting 0.5.1 ready for release. 

Now, if you go to the [Releases page]( of ArchipelaGOAL, you might notice the Linux tar package. I can't guarantee that suddenly means that the mod will _work_on Linux, but it does mean you could _try_ it. Obviously the client wouldn't work, but I would be curious to know if the mod itself, plus the REPL, would work on Linux!

**Comment by massimilianodelliubaldini** [comment]
Thanks GitHub mobile, very cool.

**Comment by tioui** [comment]
I did start working on the Linux support some time ago, but since the mod was still very intensively updated, I had the impression to actually working on things that could still change a lots in the future. So I stopped and waited. The problem is that my free time is gone for now and I don't think that I will be able to looked at it in the near future.

**Comment by massimilianodelliubaldini** [comment]
That's on me lol, I ended up iterating a ton throughout development as I used player feedback to judge what features/bugfixes I needed to add. However, Barg and Zed have both confirmed that the latest OpenGOAL mod base can build Linux binaries, and ArchipelaGOAL v0.4.0 does have them. So that's great news! That "last thing that needs to get tackled to support Linux" is basically here and now!

Unfortunately, I can't commit to fixing `MemoryReader.py` right now, nor merging in a PR for it, but if anyone does want to try their hand at implementing `PyMemoryEditor` in place of `pymem` and testing it out, I would love to see it and could test it out on Windows. We could make a separate apworld release for Linux as a stopgap, or some other compromise, but we'll cross that bridge when we get there.

**Comment by jacobmix** [comment]
Anyone following Linux for ArchipelaGOAL there's been progress: https://github.com/ArchipelaGOAL/Archipelago/pull/84
A beta test apworld can be found here: https://github.com/ArchipelaGOAL/Archipelago/releases/tag/v1.1.0
From Discord:
> ## Jak and Daxter APWORLD version 1.1.0 (Linux Support Beta)
> 
> **[APWORLD Download](https://github.com/ArchipelaGOAL/Archipelago/releases/tag/v1.1.0)**
> ### WARNING
> 
> This apworld is mostly untested. Since I don't run Linux, I cannot make any promises on how well this will work, if at all. Your mileage will vary depending on your operating system, distro, and version. Downloading this apworld means you're accepting an extra level of effort for installation and setup, and I can only provide minimal support.
> 
> ### Installation Instructions
> 
> 1. You may need to install `PyMemoryEditor v1.5.24` as a prerequisite Python package. It is not included by Archipelago.
> 
> 2. Backup your current `jakanddaxter.apworld` in `Archipelago/lib/worlds` to a safe place, preferably outside the Archipelago folder itself. You can move it, or you can copy it, **but it must also be removed from this folder.**
> 
> 3. Copy the beta `jakanddaxter.apworld` into `Archipelago/custom_worlds`.
> 
> 4. Test by generating a solo Jak and Daxter seed, hosting it on localhost, and running the Jak client. The client should be able to run the game and compiler automatically, and it should compile the game and set you up on the title screen for the game.
> 
> Please report issues [here](https://github.com/ArchipelaGOAL/Archipelago/issues), **do not report them on the main Archipelago repo.**

Some extra info:
You can get PyMemoryEditor with something like: ``py -3 -m pip install PyMemoryEditor==1.5.24 --target ./pymemedi`` then copy the ``PyMemoryEditor`` & ``pymemoryeditor-1.5.24.dist-info`` folder into ``...\Archipelago\lib``
Tho you might have to make sure you have Python 3.12. (You can get Python from [here](<https://www.python.org/downloads/release/python-31210/>) or use ``sudo apt-get update`` & ``sudo apt-get install python3.12.10``
To install update pip: ``curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py`` ``py -3 get-pip.py`` ``py -3 -m pip install --upgrade pip`` (Note: ``py -3`` might be a Windows thing. Linux might ask for ``python3`` instead.)

Also seems the apworld doesn't work on frozen. Except if you edit ``client.py`` something like:
```py
from CommonClient import ClientCommandProcessor, CommonContext, server_loop, gui_enabled
from Launcher import launch as program_launch
from NetUtils import ClientStatus
```
with:
```py
from CommonClient import ClientCommandProcessor, CommonContext, server_loop, gui_enabled
try:
    from Launcher import launch as program_launch
except ImportError:
    def program_launch(server_address=None, password=None):
        ctx = JakAndDaxterContext(server_address, password)
        import asyncio
        asyncio.run(ctx.run_gui())
from NetUtils import ClientStatus
```
*Unsanctioned* apworld download with the changes [here](https://cdn.discordapp.com/attachments/1394026388683755613/1409011633979654265/jakanddaxter.apworld?ex=68abd3d7&is=68aa8257&hm=b13762bd931ed25ad2c0e51e89fefd1f7b35e9aa7539263fb8142da8d05656c7&).
Above should fix this error:
```
Uncaught Exception:
Traceback (most recent call last):
  File "kivy\base.py", line 339, in mainloop
  File "kivy\base.py", line 383, in idle
  File "kivy\base.py", line 334, in dispatch_input
  File "kivy\base.py", line 302, in post_dispatch_input
  File "kivy\\_event.pyx", line 731, in kivy._event.EventDispatcher.dispatch
  File "kivymd\uix\behaviors\ripple_behavior.py", line 413, in on_touch_up
  File "kivy\uix\behaviors\button.py", line 179, in on_touch_up
  File "kivy\\_event.pyx", line 727, in kivy._event.EventDispatcher.dispatch
  File "kivy\\_event.pyx", line 1307, in kivy._event.EventObservers.dispatch
  File "kivy\\_event.pyx", line 1191, in kivy._event.EventObservers._dispatch
  File "kivy\lang\builder.py", line 60, in custom_callback
  File "C:\ProgramData\Archipelago\data\launcher.kv", line 58, in <module>
    on_release: app.component_action(self)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "Launcher.py", line 398, in component_action
  File "C:\ProgramData\Archipelago\custom_worlds\jakanddaxter.apworld\jakanddaxter\__init__.py", line 50, in launch_client
    from . import client
  File "C:\ProgramData\Archipelago\custom_worlds\jakanddaxter.apworld\jakanddaxter\client.py", line 24, in <module>
    from Launcher import launch as program_launch
ModuleNotFoundError: No module named 'Launcher'
```

If you do make a GitHub issue or post on Discord. Please attach as much info as you can, and logs can be found here:
> - `Archipelago\logs`
> - `OpenGOAL\features\jak1\mods\JakMods\archipelagoal\data\log`

**Comment by tioui** [comment]
On Ubuntu, the `xdg-terminal-exec` command is not installed by default. So you need to install it by doing `sudo apt install xdg-terminal-exec` .

Also @massimilianodelliubaldini , could we not just add the `PyMemoryEditor` in the `requirements.txt` file?

**Comment by massimilianodelliubaldini** [comment]
That's what I've done: [here](https://github.com/ArchipelaGOAL/Archipelago/blob/jak1-replace-mem-reader/worlds/jakanddaxter/requirements.txt). I think the issue is frozen AP builds can't/don't load libraries they don't come packaged with. So even if you installed PME yourself, your frozen AP install doesn't know about it. That's why Flit was only able to get this working by running from source.

I think that means I need to vendor PME the same way other custom apworlds do.

**Comment by FlitPix** [comment]
For anyone not following prereleases, the above issue has been fixed - specifically, PME is bundled into the APWorld now.

This is a collection of issues on 1.1.2, some I've already mentioned in the AP Discord and some I've not.

On frozen, the Jak client still doesn't open 0.6.3 without jacobmix's client.py import code. After doing that, unfortunately this is still quite unusable.
Firstly, when I launch the Jak client, the game opens, but not the REPL. I have to open the OpenGOAL Launcher, then go to Features > Mods > ArchipelaGOAL, then hit Advanced > Open REPL; then use `/repl connect` in the Jak client. This is the error thrown:
```
Traceback (most recent call last):
  File "[...]/Archipelago/custom_worlds/jakanddaxter.apworld/jakanddaxter/client.py", line 50, in _log_exception
    return await a
           ^^^^^^^
  File "[...]/Archipelago/custom_worlds/jakanddaxter.apworld/jakanddaxter/client.py", line 620, in run_game
    goalc_process = program_launch(goalc_args, in_terminal=True)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: program_launch() got an unexpected keyword argument 'in_terminal'
```
Either way, I can connect, but I cannot send checks. I don't see any obvious errors that might cause this. I haven't tested if I can receive items from other games, though I suspect this is unlikely.
Here are the logs for that: [jak1.2025-09-21T20-40-52.log](https://github.com/user-attachments/files/22455890/jak1.2025-09-21T20-40-52.log) [JakAndDaxterClient_2025_09_21_20_30_17.txt](https://github.com/user-attachments/files/22455896/JakAndDaxterClient_2025_09_21_20_30_17.txt) [JakAndDaxterGame_2025_09_21_20_30_19.txt](https://github.com/user-attachments/files/22455898/JakAndDaxterGame_2025_09_21_20_30_19.txt)

Meanwhile, running Jak client from source, the REPL opens on its own, but the Jak client freezes and "stops responding" for a good while (a minute or two?) before the game finally starts to compile. From there, though, everything seems to run beautifully. I've played up to Boggy Swamp without any issues and plan on doing a solo multiworld at some point.

TL;DR: Runs without any issues when running from AP source! Doesn't open REPL or send checks when running from AP frozen.

**Comment by massimilianodelliubaldini** [comment]
Merged into AP mainline with https://github.com/ArchipelagoMW/Archipelago/pull/5850. 

Always feels good closing a single-digit issue. :)
